File: aoc/day16.py
import re

ref = { "children": lambda x: x == 3,
        "cats": lambda x: x > 7,
        "trees": lambda x: x > 3,
        "samoyeds": lambda x: x == 2,
        "akitas": lambda x: x == 0,
        "vizslas": lambda x: x == 0,
        "pomeranians": lambda x: x < 3,
        "goldfish": lambda x: x < 5,
        "cars": lambda x: x == 2,
        "perfumes": lambda x: x == 1 }

with open('input.txt', 'r') as fh:
    p = re.compile(r'^Sue ([0-9]+): ([A-Za-z]+): ([0-9]+), ([A-Za-z]+): ([0-9]+), ([A-Za-z]+): ([0-9]+)$')
    for l in fh:
        match = p.findall(l.strip())[0]
        nr = match[0]
        things = dict(zip([name for i, name in enumerate(match[1:]) if i % 2 == 0],
                          [int(count) for i, count in enumerate(match[1:]) if i % 2 == 1]))

        if sum([ref[k](v) and 1 or 0 for k,v in things.items()]) == 3: print (nr, things)

File: aoc/Day16/src/Day16.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class Day16 {

	public static void main(String[] args) throws FileNotFoundException {
		
		Scanner input = new Scanner(new File("input.txt"));
		int curr = 0;
		String in;
		int num;
		boolean v1 = false, v2 = false, v3 = false;
		
		for (int i = 0; i < 500; i++) {
			input.next();
			input.next();
			
			in = input.next();
			num = Integer.parseInt(input.next().replace(',', ' ').trim());
			System.out.println(in + " " + num);
			v1 = valid(in, num);
			
			in = input.next();
			num = Integer.parseInt(input.next().replace(',', ' ').trim());
			System.out.println(in + " " + num);
			v2 = valid(in, num);
			
			in = input.next();
			num = Integer.parseInt(input.next().replace(',', ' ').trim());
			System.out.println(in + " " + num);
			v3 = valid(in, num);
			
			if (v1 && v2 && v3)
				curr = i;
			
			
		}
		
		input.close();
		
	}
	
	public static boolean valid(String in, int num) {
		boolean isValid = false;
		int[] Sue = {3, 7, 2, 3, 0, 0, 5, 3, 2, 1};
		int i;
		
		if (in == "children:")
			i = 0;
		else if (in == "cats:")
			i = 1;
		else if (in == "samoyeds:")
			i = 2;
		else if (in == "pomeranians:")
			i = 3;
		else if (in == "akitas:")
			i = 4;
		else if (in == "vizslas:")
			i = 5;
		else if (in == "goldfish:")
			i = 6;
		else if (in == "trees:")
			i = 7;
		else if (in == "cars:")
			i = 8;
		else if (in == "perfumes:")
			i = 9;
		else
			i = -1;
			
		if (i != -1 && Sue[i] == num)
			isValid = true;
		
		return isValid;
	}

}
